File: MCTS_TicTacToe_python/Board.py
import numpy as np
from Position import Position


def check_for_win(row):
    is_equal = True
    size = len(row)
    previous = row[0]
    for i in range(1, size):
        if previous != row[i]:
            is_equal = False
            break
        previous = row[i]

    if is_equal:
        return previous
    else:
        return 0


class Board:
    DEFAULT_BOARD_SIZE = 3
    IN_PROGRESS = -1
    DRAW = 0
    P1 = 1
    P2 = 2

    def __init__(self):
        self.totalMoves = 0
        self.boardValues = np.zeros((3, 3))

    def __str__(self):
        board = ""
        size = len(self.boardValues)
        for i in range(0, size):
            for j in range(0, size):
                board += str(int(self.boardValues[i, j])) + " "
            board += "\n"

        return board

    def perform_move(self, player, position):
        self.totalMoves += 1
        self.boardValues[position.x, position.y] = player

    def check_status(self):
        board_size = len(self.boardValues)
        max_index = board_size - 1
        diag1 = np.empty(board_size)
        diag2 = np.empty(board_size)

        for i in range(0, board_size):
            row = self.boardValues[i, :]
            col = np.empty(board_size)
            for j in range(0, board_size):
                col[j] = self.boardValues[i, j]

            check_row_for_win = check_for_win(row)
            if check_row_for_win != 0:
                return check_row_for_win

            check_col_for_win = check_for_win(col)
            if check_col_for_win != 0:
                return check_col_for_win

            diag1[i] = self.boardValues[i, i]
            diag2[i] = self.boardValues[max_index - i, i]

        check_diag1_for_win = check_for_win(diag1)
        if check_diag1_for_win != 0:
            return check_diag1_for_win

        check_diag2_for_win = check_for_win(diag2)
        if check_diag2_for_win != 0:
            return check_diag2_for_win

        if len(self.get_empty_positions()) > 0:
            return Board.IN_PROGRESS
        else:
            return Board.DRAW

    def get_empty_positions(self):
        size = len(self.boardValues)
        empty_positions = []
        for i in range(0, size):
            for j in range(0, size):
                if self.boardValues[i, j] == 0:
                    new_position = Position(i, j)
                    empty_positions.append(new_position)
        return empty_positions

File: MCTS_TicTacToe_cpp/Board.cpp
//
// Created by João Vieira on 22/12/2019.
//

#include <iostream>
#include "Board.h"
#include "Position.h"

Board::Board() {
    boardValues = new int*[DEFAULT_BOARD_SIZE];
    for (int i = 0; i < DEFAULT_BOARD_SIZE; ++i) {
        boardValues[i] = new int[DEFAULT_BOARD_SIZE];

        for (int j = 0; j < DEFAULT_BOARD_SIZE; ++j) {
            boardValues[i][j] = 0;
        }
    }

    totalMoves = 0;
}

int Board::checkStatus() {
    int maxIndex = DEFAULT_BOARD_SIZE - 1;
    int diag1[DEFAULT_BOARD_SIZE], diag2[DEFAULT_BOARD_SIZE];

    for (int i = 0; i < DEFAULT_BOARD_SIZE; i++) {
        int *row = boardValues[i];
        int col[DEFAULT_BOARD_SIZE];
        for (int j = 0; j < DEFAULT_BOARD_SIZE; j++) {
            col[j] = boardValues[j][i];
        }

        int checkRowForWin = checkForWin(row);
        if(checkRowForWin!=0)
            return checkRowForWin;

        int checkColForWin = checkForWin(col);
        if(checkColForWin!=0)
            return checkColForWin;

        diag1[i] = boardValues[i][i];
        diag2[i] = boardValues[maxIndex - i][i];
    }

    int checkDia1gForWin = checkForWin(diag1);
    if(checkDia1gForWin!=0)
        return checkDia1gForWin;

    int checkDiag2ForWin = checkForWin(diag2);
    if(checkDiag2ForWin!=0)
        return checkDiag2ForWin;

    if (!getEmptyPositions()->empty())
        return IN_PROGRESS;
    else
        return DRAW;
}

Board::Board(Board *board) {
    boardValues = new int*[DEFAULT_BOARD_SIZE];
    for (int i = 0; i < DEFAULT_BOARD_SIZE; ++i) {
        boardValues[i] = new int[DEFAULT_BOARD_SIZE];

        for (int j = 0; j < DEFAULT_BOARD_SIZE; ++j) {
            boardValues[i][j] = 0;
        }
    }

    for (int i = 0; i < DEFAULT_BOARD_SIZE; ++i) {
        for (int j = 0; j < DEFAULT_BOARD_SIZE; ++j) {
            boardValues[i][j] = board->getBoardValues()[i][j];
        }
    }

    totalMoves = board->getTotalMoves();
}

int **Board::getBoardValues() {
    return boardValues;
}

int Board::checkForWin(const int *row) {
    bool isEqual = true;
    int previous = row[0];
    for (int i = 0; i < DEFAULT_BOARD_SIZE; i++) {
        if (previous != row[i]) {
            isEqual = false;
            break;
        }
        previous = row[i];
    }
    if(isEqual)
        return previous;
    else
        return 0;
}

std::list<Position *> *Board::getEmptyPositions() {
    auto emptyPositions = new std::list<Position *>();

    for (int i = 0; i < DEFAULT_BOARD_SIZE; ++i) {
        for (int j = 0; j < DEFAULT_BOARD_SIZE; ++j) {
            if (!boardValues[i][j])
                emptyPositions->push_back(new Position(i, j));
        }
    }

    return emptyPositions;
}

void Board::performMove(int player, Position *p) {
    totalMoves++;
    boardValues[p->getX()][p->getY()] = player;
}

int Board::getTotalMoves() {
    return totalMoves;
}

void Board::printBoard() {
    for (int i = 0; i < DEFAULT_BOARD_SIZE; ++i) {
        for (int j = 0; j < DEFAULT_BOARD_SIZE; ++j) {
            std::cout << boardValues[i][j] << " ";
        }

        std::cout << "\n";
    }
}

void Board::printStatus() {
    switch (checkStatus()) {
        case P1:
            std::cout << "Player 1 wins\n";
            break;
        case P2:
            std::cout << "Player 2 wins\n";
            break;
        case DRAW:
            std::cout << "Game Draw\n";
            break;
        default:
            std::cout << "Game In Progress\n";
            break;
    }
}
